In [1]:
import pandas as pd
import numpy as np
import warnings
from collections import Counter
import time
warnings.filterwarnings('ignore')

import sys
sys.path.append('..')
from src.utils.data_loader import load_paysim_data

print("="*80)
print("CLEAN FEATURE ENGINEERING - NO LEAKAGE")

# Load data
print("Loading data and creating time-based split...")
df = load_paysim_data()

# Add time columns
df['day'] = df['step'] // 24
df['hour_of_day'] = df['step'] % 24

# Time-based split (last 7 days = test)
test_days = 7
test_steps = test_days * 24
train_threshold = df['step'].max() - test_steps

print(f"\nData Configuration:")
print(f"  Total transactions: {len(df):,}")
print(f"  Train period: steps 0-{train_threshold} ({train_threshold/24:.1f} days)")
print(f"  Test period: steps {train_threshold+1}-{df['step'].max()} ({test_days} days)")

# Create train/test dataframes
train_mask = df['step'] <= train_threshold
train_df = df[train_mask].copy()
test_df = df[~train_mask].copy()

print(f"\nSplit Statistics:")
print(f"  Train: {len(train_df):,} samples, {train_df['isFraud'].sum():,} frauds ({train_df['isFraud'].mean():.3%})")
print(f"  Test: {len(test_df):,} samples, {test_df['isFraud'].sum():,} frauds ({test_df['isFraud'].mean():.3%})")
print(f"  Concept drift: {test_df['isFraud'].mean() / train_df['isFraud'].mean():.1f}x fraud rate increase")

print("\n✓ Clean data split complete")

CLEAN FEATURE ENGINEERING - NO LEAKAGE
Loading data and creating time-based split...
Dataset loaded successfully!
Shape: (6362620, 11)
Memory usage: 260.92 MB

Fraud rate: 0.1291%
Flagged fraud rate: 0.0003%

Data Configuration:
  Total transactions: 6,362,620
  Train period: steps 0-575 (24.0 days)
  Test period: steps 576-743 (7 days)

Split Statistics:
  Train: 6,200,317 samples, 6,359 frauds (0.103%)
  Test: 162,303 samples, 1,854 frauds (1.142%)
  Concept drift: 11.1x fraud rate increase

✓ Clean data split complete


In [2]:
def create_truly_clean_features(df, train_df):
    """
    Features with ABSOLUTE ZERO usage of balance columns
    Only: amount, type, time - nothing from balances
    """
    start_time = time.time()
    
    # === TEMPORAL FEATURES ===
    print("  [1/3] Temporal features...")
    
    df['hour_sin'] = np.sin(2 * np.pi * df['hour_of_day'] / 24)
    df['hour_cos'] = np.cos(2 * np.pi * df['hour_of_day'] / 24)
    df['day_sin'] = np.sin(2 * np.pi * df['day'] / 30)
    df['day_cos'] = np.cos(2 * np.pi * df['day'] / 30)
    
    # Fraud peak hours (146x lift verified)
    df['is_fraud_peak_hour'] = ((df['hour_of_day'] >= 3) & (df['hour_of_day'] <= 6)).astype(int)
    df['is_night'] = ((df['hour_of_day'] >= 22) | (df['hour_of_day'] <= 6)).astype(int)
    df['is_business_hours'] = ((df['hour_of_day'] >= 9) & (df['hour_of_day'] <= 17)).astype(int)
    
    # === AMOUNT FEATURES (NO balance ratios) ===
    print("  [2/3] Amount features...")
    
    df['log_amount'] = np.log10(df['amount'] + 1)
    
    # Amount by type (from training only)
    type_stats = train_df.groupby('type')['amount'].agg(['mean', 'std']).reset_index()
    type_stats.columns = ['type', 'type_mean_amount', 'type_std_amount']
    
    type_mean_map = type_stats.set_index('type')['type_mean_amount'].to_dict()
    type_std_map = type_stats.set_index('type')['type_std_amount'].to_dict()
    
    df['type_mean_amount'] = df['type'].astype(str).map(type_mean_map).astype(float)
    df['type_std_amount'] = df['type'].astype(str).map(type_std_map).astype(float)
    
    df['amount_zscore_by_type'] = (
        (df['amount'] - df['type_mean_amount']) / (df['type_std_amount'] + 1e-6)
    )
    
    # High amount flags
    df['is_high_amount'] = (df['amount'] > 1000000).astype(int)
    df['is_round_thousand'] = (df['amount'] % 1000 == 0).astype(int)
    df['is_round_10k'] = (df['amount'] % 10000 == 0).astype(int)
    
    # === BEHAVIORAL FEATURES ===
    print("  [3/3] Behavioral features...")
    
    df['is_risky_type'] = df['type'].isin(['TRANSFER', 'CASH_OUT']).astype(int)
    df['is_transfer'] = (df['type'] == 'TRANSFER').astype(int)
    df['is_cashout'] = (df['type'] == 'CASH_OUT').astype(int)
    
    # === INTERACTIONS (681x lift verified) ===
    df['risky_type_fraud_hour'] = df['is_risky_type'] * df['is_fraud_peak_hour']
    df['risky_type_high_amount'] = df['is_risky_type'] * df['is_high_amount']
    df['high_amount_night'] = df['is_high_amount'] * df['is_night']
    
    # REMOVED: empties_origin_account, dest_starts_zero, amount_to_orig_balance
    # These all use balance columns
    
    elapsed = time.time() - start_time
    print(f"  Completed in {elapsed:.1f}s")
    
    return df

# Apply truly clean features
df = create_truly_clean_features(df, train_df)

# Validate
print("\n  Feature validation:")
key_features = ['is_fraud_peak_hour', 'risky_type_fraud_hour', 'is_high_amount', 'is_round_thousand']

for col in key_features:
    fraud_mean = df[df['isFraud']==1][col].mean()
    normal_mean = df[df['isFraud']==0][col].mean()
    if normal_mean > 0:
        ratio = fraud_mean / normal_mean
        print(f"    {col}: Fraud={fraud_mean:.3f}, Normal={normal_mean:.3f}, Ratio={ratio:.1f}x")

print("\n✓ Truly clean features (ZERO balance column usage)")

  [1/3] Temporal features...
  [2/3] Amount features...
  [3/3] Behavioral features...
  Completed in 2.0s

  Feature validation:
    is_fraud_peak_hour: Fraud=0.161, Normal=0.001, Ratio=146.7x
    risky_type_fraud_hour: Fraud=0.161, Normal=0.000, Ratio=681.1x
    is_high_amount: Fraud=0.329, Normal=0.020, Ratio=16.4x
    is_round_thousand: Fraud=0.037, Normal=0.001, Ratio=70.2x

✓ Truly clean features (ZERO balance column usage)


In [3]:
def prepare_clean_feature_matrix(df):
    """
    Create feature matrix excluding forbidden columns
    """
    
    # === FORBIDDEN COLUMNS ===
    forbidden = [
        # Balance columns (case study requirement)
        'oldbalanceOrg', 'newbalanceOrig', 'oldbalanceDest', 'newbalanceDest',
        # Intermediate calculation columns
        'type_mean_amount', 'type_std_amount',
        # Identifiers
        'nameOrig', 'nameDest', 'step', 'day', 'hour_of_day',
        # Labels
        'isFraud', 'isFlaggedFraud',
        # Raw type (will use one-hot)
        'type'
    ]
    
    forbidden = [col for col in forbidden if col in df.columns]
    
    # Get feature columns
    feature_cols = [col for col in df.columns if col not in forbidden]
    
    # One-hot encode transaction type
    type_dummies = pd.get_dummies(df['type'], prefix='type', drop_first=False)
    
    # Create feature matrix
    X = pd.concat([df[feature_cols], type_dummies], axis=1)
    y = df['isFraud']
    
    final_feature_names = list(feature_cols) + list(type_dummies.columns)
    
    return X, y, final_feature_names

# Prepare features
X, y, feature_names = prepare_clean_feature_matrix(df)

print(f"  Feature matrix: {X.shape}")
print(f"  Total features: {len(feature_names)}")

# Create train/test split
print(f"\n  Creating train/test split (temporal: step <= {train_threshold})...")

X_train = X[df['step'] <= train_threshold].copy()
X_test = X[df['step'] > train_threshold].copy()
y_train = y[df['step'] <= train_threshold].copy()
y_test = y[df['step'] > train_threshold].copy()

print(f"\n  Train: {len(X_train):,} samples, {y_train.sum():,} frauds ({y_train.mean():.3%})")
print(f"  Test: {len(X_test):,} samples, {y_test.sum():,} frauds ({y_test.mean():.3%})")
print(f"  Class imbalance: 1:{int((y_train==0).sum() / (y_train==1).sum())}")

# Show top features
print(f"\n  Top 10 features by signal strength:")
feature_signals = []
for col in feature_names:
    if X[col].std() > 0:
        fraud_mean = X[y == 1][col].mean()
        normal_mean = X[y == 0][col].mean()
        if normal_mean != 0:
            ratio = abs(fraud_mean / normal_mean)
            feature_signals.append((col, ratio))

feature_signals.sort(key=lambda x: x[1], reverse=True)
for i, (feat, ratio) in enumerate(feature_signals[:10], 1):
    print(f"    {i}. {feat}: {ratio:.1f}x")

print("\n✓ Clean feature matrix ready")

  Feature matrix: (6362620, 24)
  Total features: 24

  Creating train/test split (temporal: step <= 575)...

  Train: 6,200,317 samples, 6,359 frauds (0.103%)
  Test: 162,303 samples, 1,854 frauds (1.142%)
  Class imbalance: 1:974

  Top 10 features by signal strength:
    1. risky_type_fraud_hour: 681.1x
    2. amount_zscore_by_type: 653.2x
    3. is_fraud_peak_hour: 146.7x
    4. high_amount_night: 109.5x
    5. is_round_10k: 75.8x
    6. is_round_thousand: 70.2x
    7. risky_type_high_amount: 16.4x
    8. is_high_amount: 16.4x
    9. amount: 8.2x
    10. is_transfer: 6.0x

✓ Clean feature matrix ready


## This is a pre baseline model!

In [4]:
import xgboost as xgb
from sklearn.metrics import (
    average_precision_score, roc_auc_score,
    precision_score, recall_score, f1_score,
    confusion_matrix
)

# Calculate class weight
scale_pos_weight = (y_train == 0).sum() / (y_train == 1).sum()
print(f"Class imbalance: 1:{scale_pos_weight:.0f}")

# Train XGBoost with cost-sensitive learning
print(f"\nTraining XGBoost on {len(X_train):,} samples...")
start_time = time.time()

model = xgb.XGBClassifier(
    n_estimators=100,
    max_depth=6,
    learning_rate=0.1,
    scale_pos_weight=scale_pos_weight,  # Cost-sensitive
    random_state=42,
    tree_method='hist',
    n_jobs=-1
)

model.fit(X_train, y_train, verbose=False)
train_time = time.time() - start_time

print(f"Training completed in {train_time:.1f}s")

# Predictions
y_pred_proba_train = model.predict_proba(X_train)[:, 1]
y_pred_proba_test = model.predict_proba(X_test)[:, 1]

# Metrics
auprc_train = average_precision_score(y_train, y_pred_proba_train)
auprc_test = average_precision_score(y_test, y_pred_proba_test)
auc_test = roc_auc_score(y_test, y_pred_proba_test)

print(f"\nPerformance:")
print(f"  Train AUPRC: {auprc_train:.3f}")
print(f"  Test AUPRC: {auprc_test:.3f}")
print(f"  Test AUC-ROC: {auc_test:.3f}")

# Check overfitting
if auprc_train > auprc_test * 1.5:
    print(f"\n⚠️ Overfitting detected (train/test ratio: {auprc_train/auprc_test:.1f}x)")
else:
    print(f"\n✓ Reasonable generalization (train/test ratio: {auprc_train/auprc_test:.1f}x)")

# Threshold optimization (Document 8: FN=$500, FP=$5)
print(f"\nOptimizing threshold (FN=$500, FP=$5)...")

def calculate_cost(y_true, y_pred):
    fp = ((y_pred == 1) & (y_true == 0)).sum()
    fn = ((y_pred == 0) & (y_true == 1)).sum()
    return 5 * fp + 500 * fn

thresholds = np.arange(0.001, 0.5, 0.001)
costs = [calculate_cost(y_test, (y_pred_proba_test >= t).astype(int)) for t in thresholds]
optimal_idx = np.argmin(costs)
optimal_threshold = thresholds[optimal_idx]
optimal_cost = costs[optimal_idx]

y_pred_optimal = (y_pred_proba_test >= optimal_threshold).astype(int)

print(f"  Optimal threshold: {optimal_threshold:.4f}")
print(f"  Minimum cost: ${optimal_cost:,}")

# Classification metrics at optimal threshold
precision = precision_score(y_test, y_pred_optimal, zero_division=0)
recall = recall_score(y_test, y_pred_optimal, zero_division=0)
f1 = f1_score(y_test, y_pred_optimal, zero_division=0)

print(f"\nClassification Metrics (at optimal threshold):")
print(f"  Precision: {precision:.3f}")
print(f"  Recall: {recall:.3f}")
print(f"  F1-Score: {f1:.3f}")

# Confusion matrix
cm = confusion_matrix(y_test, y_pred_optimal)
print(f"\nConfusion Matrix:")
print(f"  TN: {cm[0,0]:,} | FP: {cm[0,1]:,}")
print(f"  FN: {cm[1,0]:,} | TP: {cm[1,1]:,}")

# Operational metrics
if y_pred_optimal.sum() > 0:
    sar_conversion = cm[1,1] / y_pred_optimal.sum() * 100
    print(f"\nOperational Metrics:")
    print(f"  Total alerts: {y_pred_optimal.sum():,}")
    print(f"  SAR conversion: {sar_conversion:.1f}%")
    print(f"  Frauds caught: {cm[1,1]:,}/{y_test.sum():,} ({recall*100:.1f}%)")
    print(f"  False positive rate: {cm[0,1]/len(y_test)*100:.2f}%")
    
    # Document 3 benchmarks
    print(f"\n  Document 3 Benchmarks:")
    print(f"    Basic ML SAR conversion: 5-10%")
    print(f"    Advanced ML: 10-15%")
    print(f"    Our result: {sar_conversion:.1f}%")
else:
    print("\n⚠️ Model flags zero transactions")

print("\n✓ Baseline complete")

Class imbalance: 1:974

Training XGBoost on 6,200,317 samples...
Training completed in 4.7s

Performance:
  Train AUPRC: 0.484
  Test AUPRC: 0.341
  Test AUC-ROC: 0.893

✓ Reasonable generalization (train/test ratio: 1.4x)

Optimizing threshold (FN=$500, FP=$5)...
  Optimal threshold: 0.0010
  Minimum cost: $331,315

Classification Metrics (at optimal threshold):
  Precision: 0.027
  Recall: 1.000
  F1-Score: 0.053

Confusion Matrix:
  TN: 94,186 | FP: 66,263
  FN: 0 | TP: 1,854

Operational Metrics:
  Total alerts: 68,117
  SAR conversion: 2.7%
  Frauds caught: 1,854/1,854 (100.0%)
  False positive rate: 40.83%

  Document 3 Benchmarks:
    Basic ML SAR conversion: 5-10%
    Advanced ML: 10-15%
    Our result: 2.7%

✓ Baseline complete


- Test AUPRC: 0.988 (near perfect)
- Test AUC-ROC: 1.000 (literally perfect)
- Recall: 99.9% (caught 1,852 of 1,854 frauds)
- Precision: 89.6%

- This is a classic case of data leakage, or The test set is too easy, or paysims synthetic nature, or i am incredibly lucky.

In [7]:
print("="*80)
print("THRESHOLD OPTIMIZATION FIX")
print("="*80)

# Current issue: Threshold 0.001 flags 68k transactions
print(f"\nCurrent (broken) threshold: 0.001")
print(f"  Alerts: 68,117")
print(f"  SAR conversion: 2.7%")
print(f"  Problem: Flagging 42% of transactions is operationally impossible\n")

# Try different threshold search ranges
print("Testing threshold ranges:")

# Range 1: Higher thresholds
thresholds_high = np.arange(0.01, 0.99, 0.01)
costs_high = []
precisions_high = []
recalls_high = []

for t in thresholds_high:
    y_pred = (y_pred_proba_test >= t).astype(int)
    fp = ((y_pred == 1) & (y_test == 0)).sum()
    fn = ((y_pred == 0) & (y_test == 1)).sum()
    cost = 5 * fp + 500 * fn
    costs_high.append(cost)
    precisions_high.append(precision_score(y_test, y_pred, zero_division=0))
    recalls_high.append(recall_score(y_test, y_pred, zero_division=0))

optimal_idx_high = np.argmin(costs_high)
optimal_threshold_high = thresholds_high[optimal_idx_high]

print(f"\n1. Higher range (0.01-0.99):")
print(f"   Optimal threshold: {optimal_threshold_high:.3f}")
print(f"   Min cost: ${costs_high[optimal_idx_high]:,}")

# Evaluate at this threshold
y_pred_fixed = (y_pred_proba_test >= optimal_threshold_high).astype(int)

precision_fixed = precision_score(y_test, y_pred_fixed, zero_division=0)
recall_fixed = recall_score(y_test, y_pred_fixed, zero_division=0)
f1_fixed = f1_score(y_test, y_pred_fixed, zero_division=0)

cm_fixed = confusion_matrix(y_test, y_pred_fixed)

print(f"\nFixed Threshold Performance:")
print(f"  Precision: {precision_fixed:.3f}")
print(f"  Recall: {recall_fixed:.3f}")
print(f"  F1-Score: {f1_fixed:.3f}")

print(f"\nConfusion Matrix:")
print(f"  TN: {cm_fixed[0,0]:,} | FP: {cm_fixed[0,1]:,}")
print(f"  FN: {cm_fixed[1,0]:,} | TP: {cm_fixed[1,1]:,}")

if y_pred_fixed.sum() > 0:
    sar_conv_fixed = cm_fixed[1,1] / y_pred_fixed.sum() * 100
    print(f"\nOperational Metrics:")
    print(f"  Total alerts: {y_pred_fixed.sum():,}")
    print(f"  SAR conversion: {sar_conv_fixed:.1f}%")
    print(f"  Frauds caught: {cm_fixed[1,1]:,}/{y_test.sum():,} ({recall_fixed*100:.1f}%)")
    
    print(f"\n  Document 3 Benchmarks:")
    print(f"    Basic ML: 5-10% SAR conversion")
    print(f"    Advanced ML: 10-15%")
    print(f"    Our result: {sar_conv_fixed:.1f}%")
else:
    print("\n⚠️ No alerts at this threshold")

print(f"\n✓ Threshold optimization fixed")

THRESHOLD OPTIMIZATION FIX

Current (broken) threshold: 0.001
  Alerts: 68,117
  SAR conversion: 2.7%
  Problem: Flagging 42% of transactions is operationally impossible

Testing threshold ranges:

1. Higher range (0.01-0.99):
   Optimal threshold: 0.010
   Min cost: $331,315

Fixed Threshold Performance:
  Precision: 0.027
  Recall: 1.000
  F1-Score: 0.053

Confusion Matrix:
  TN: 94,186 | FP: 66,263
  FN: 0 | TP: 1,854

Operational Metrics:
  Total alerts: 68,117
  SAR conversion: 2.7%
  Frauds caught: 1,854/1,854 (100.0%)

  Document 3 Benchmarks:
    Basic ML: 5-10% SAR conversion
    Advanced ML: 10-15%
    Our result: 2.7%

✓ Threshold optimization fixed


In [8]:
print("="*80)
print("BUSINESS-CONSTRAINED THRESHOLD OPTIMIZATION")
print("="*80)

print("\nScenario 1: Alert Budget Constraint")
print("-" * 50)
print("Realistic operational capacity: 1-5% of transactions can be reviewed")

# Set alert budget: 2% of test transactions
alert_budget = int(len(y_test) * 0.02)  # 2% = ~3,246 alerts

print(f"\nAlert budget: {alert_budget:,} (2% of {len(y_test):,} transactions)")

# Find threshold that produces this many alerts
sorted_proba = np.sort(y_pred_proba_test)[::-1]  # Descending
threshold_budget = sorted_proba[alert_budget] if alert_budget < len(sorted_proba) else 0.001

y_pred_budget = (y_pred_proba_test >= threshold_budget).astype(int)

cm_budget = confusion_matrix(y_test, y_pred_budget)
precision_budget = precision_score(y_test, y_pred_budget, zero_division=0)
recall_budget = recall_score(y_test, y_pred_budget, zero_division=0)

print(f"\nThreshold for {alert_budget:,} alerts: {threshold_budget:.4f}")
print(f"  Precision: {precision_budget:.3f}")
print(f"  Recall: {recall_budget:.3f}")
print(f"  Confusion Matrix:")
print(f"    TN: {cm_budget[0,0]:,} | FP: {cm_budget[0,1]:,}")
print(f"    FN: {cm_budget[1,0]:,} | TP: {cm_budget[1,1]:,}")

if y_pred_budget.sum() > 0:
    sar_conv_budget = cm_budget[1,1] / y_pred_budget.sum() * 100
    print(f"  SAR conversion: {sar_conv_budget:.1f}%")
    print(f"  Frauds caught: {cm_budget[1,1]:,}/{y_test.sum():,} ({recall_budget*100:.1f}%)")

print("\n" + "-"*50)
print("Scenario 2: Minimum Recall Constraint")
print("-" * 50)
print("Requirement: Catch at least 80% of frauds")

# Find threshold for 80% recall
target_recall = 0.80
thresholds_recall = np.arange(0.001, 0.99, 0.001)

for t in thresholds_recall:
    y_pred = (y_pred_proba_test >= t).astype(int)
    recall = recall_score(y_test, y_pred, zero_division=0)
    if recall >= target_recall:
        threshold_recall = t
        y_pred_recall = y_pred
        break

cm_recall = confusion_matrix(y_test, y_pred_recall)
precision_recall = precision_score(y_test, y_pred_recall, zero_division=0)
actual_recall = recall_score(y_test, y_pred_recall, zero_division=0)

print(f"\nThreshold for {target_recall*100:.0f}% recall: {threshold_recall:.4f}")
print(f"  Precision: {precision_recall:.3f}")
print(f"  Recall: {actual_recall:.3f}")
print(f"  Total alerts: {y_pred_recall.sum():,}")
print(f"  Confusion Matrix:")
print(f"    TN: {cm_recall[0,0]:,} | FP: {cm_recall[0,1]:,}")
print(f"    FN: {cm_recall[1,0]:,} | TP: {cm_recall[1,1]:,}")

if y_pred_recall.sum() > 0:
    sar_conv_recall = cm_recall[1,1] / y_pred_recall.sum() * 100
    print(f"  SAR conversion: {sar_conv_recall:.1f}%")

print("\n" + "="*80)
print("SUMMARY")
print("="*80)
print(f"\nModel Performance: AUPRC {auprc_test:.3f}, AUC-ROC {auc_test:.3f}")
print(f"\nThreshold Strategies:")
print(f"1. Cost-optimized (FN=$500, FP=$5): 68,117 alerts, 2.7% SAR conv, 100% recall")
print(f"2. Alert budget (2% capacity): {y_pred_budget.sum():,} alerts, {sar_conv_budget:.1f}% SAR conv, {recall_budget*100:.0f}% recall")
print(f"3. Minimum recall (80%): {y_pred_recall.sum():,} alerts, {sar_conv_recall:.1f}% SAR conv, {actual_recall*100:.0f}% recall")
print(f"\nRecommendation: Use alert budget or minimum recall constraint for operational feasibility")

BUSINESS-CONSTRAINED THRESHOLD OPTIMIZATION

Scenario 1: Alert Budget Constraint
--------------------------------------------------
Realistic operational capacity: 1-5% of transactions can be reviewed

Alert budget: 3,246 (2% of 162,303 transactions)

Threshold for 3,246 alerts: 0.9902
  Precision: 0.241
  Recall: 0.422
  Confusion Matrix:
    TN: 157,984 | FP: 2,465
    FN: 1,071 | TP: 783
  SAR conversion: 24.1%
  Frauds caught: 783/1,854 (42.2%)

--------------------------------------------------
Scenario 2: Minimum Recall Constraint
--------------------------------------------------
Requirement: Catch at least 80% of frauds

Threshold for 80% recall: 0.0010
  Precision: 0.027
  Recall: 1.000
  Total alerts: 68,117
  Confusion Matrix:
    TN: 94,186 | FP: 66,263
    FN: 0 | TP: 1,854
  SAR conversion: 2.7%

SUMMARY

Model Performance: AUPRC 0.341, AUC-ROC 0.893

Threshold Strategies:
1. Cost-optimized (FN=$500, FP=$5): 68,117 alerts, 2.7% SAR conv, 100% recall
2. Alert budget (2% ca

In [9]:
print("="*80)
print("EXPORTING CLEAN DATASET FOR MODEL BUILDING")
print("="*80)

import pickle
import os

# Create output directory
os.makedirs('../data/processed', exist_ok=True)

# Package everything needed for modeling
dataset_package = {
    # Feature matrices
    'X_train': X_train,
    'X_test': X_test,
    'y_train': y_train,
    'y_test': y_test,
    
    # Metadata
    'feature_names': feature_names,
    'train_threshold': train_threshold,
    
    # Key statistics
    'metadata': {
        'total_samples': len(df),
        'train_samples': len(X_train),
        'test_samples': len(X_test),
        'train_fraud_rate': y_train.mean(),
        'test_fraud_rate': y_test.mean(),
        'concept_drift_ratio': y_test.mean() / y_train.mean(),
        'n_features': len(feature_names),
        'imbalance_ratio': (y_train == 0).sum() / (y_train == 1).sum(),
        
        # Baseline performance (for comparison)
        'baseline_xgboost': {
            'auprc_train': auprc_train,
            'auprc_test': auprc_test,
            'auc_roc_test': auc_test,
            'alert_budget_threshold': threshold_budget,
            'alert_budget_sar_conversion': sar_conv_budget,
            'alert_budget_recall': recall_budget
        }
    }
}

# Save
output_file = '../data/processed/clean_features_final.pkl'
with open(output_file, 'wb') as f:
    pickle.dump(dataset_package, f)

EXPORTING CLEAN DATASET FOR MODEL BUILDING
